vm/src/types/program.rs

```diff
@@ -83,12 +85,39 @@ impl Program {
                 constants.insert(key.clone(), value);
             }
         }
+
+        let mut hints: Vec<_> = hints.into_iter().filter(|(_, v)| !v.is_empty()).collect();
+        hints.sort_unstable_by(|x, y| x.0.cmp(&y.0));
+
+        let hints_ranges_iter = hints
+            .iter()
+            .map(|(k, v)| (k, v.len()))
+            .scan(0, |s, (k, v)| {
+                let res = (
+                    k,
+                    (
+                        *s,
+                        NonZeroUsize::new(v).expect("empty vecs already filtered"),
+                    ),
+                );
+                *s += v;
+                Some(res)
+            });
+
+        let mut hints_ranges = vec![None; hints.last().map(|(x, _)| x + 1).unwrap_or(0)];
+        for (pc, r) in hints_ranges_iter {
+            hints_ranges[*pc] = Some(r);
+        }
+
+        let hints = hints.into_iter().flat_map(|(_, v)| v).collect();
```

Comment:
I think we should extract this part into a helper function

Reply:
Hehehe I had a TODO I removed because I considered it for a later refactor.
Will do now then.